core/src/main/scala/kafka/server/ReplicationQuotaManager.scala

```diff
@@ -0,0 +1,219 @@
+
+object ReplicationQuotaManagerConfig {
+  val QuotaBytesPerSecondDefault = Long.MaxValue
+  // Always have 10 whole windows + 1 current window
+  val DefaultNumQuotaSamples = 11
+  val DefaultQuotaWindowSizeSeconds = 1
+  // Purge sensors after 1 hour of inactivity
+  val InactiveSensorExpirationTimeSeconds = 3600
+}
+
+trait ReadOnlyQuota {
+  def bound(): Int
+
+  def isThrottled(topicAndPartition: TopicAndPartition): Boolean
+
+  def isQuotaExceededBy(bytes: Int): Boolean
+}
+
+object Constants {
+  val allReplicas = Seq[Int](-1)
+}
+
+/**
+  * Tracks replication metrics and comparing them to any quotas for throttled partitions.
+  *
+  * @param config          The quota configs
+  * @param metrics         The Metrics instance
+  * @param replicationType The name / key for this quota manager, typically leader or follower
+  * @param time            Time object to use
+  */
+class ReplicationQuotaManager(val config: ReplicationQuotaManagerConfig,
+                              private val metrics: Metrics,
+                              private val replicationType: String,
+                              private val time: Time) extends Logging with ReadOnlyQuota {
+
+  private val lock = new ReentrantReadWriteLock()
+  private val throttledPartitions = new ConcurrentHashMap[String, Seq[Int]]()
+  private var quota: Quota = null
+  private val sensor = createSensor(quota, replicationType)
+
+  private def rateMetricName(): MetricName = metrics.metricName("byte-rate", replicationType, s"Tracking byte-rate for $replicationType")
+
+  /**
+    * Update the quota
+    *
+    * @param quota
+    */
+  def updateQuota(quota: Quota) = {
+    inWriteLock(lock) {
+      this.quota = quota
+      if (metrics.metrics.containsKey(rateMetricName))
+        metrics.metrics.get(rateMetricName).config(getQuotaMetricConfig(quota))
+    }
+  }
+
+  /**
+    * Retrieve the current quota object
+    *
+    * @return
+    */
+  def getQuota(): Quota = {
+    inReadLock(lock) {
+      quota
+    }
+  }
+
+  /**
+    * Check if the quota is currently exceeded
+    *
+    * @return
+    */
+  def isQuotaExceeded(): Boolean = isQuotaExceededBy(0)
+
+  /**
+    * Check if the quota will be exceeded by the passed value. This method does not record anything, but is equivalent to
+    * increasing the quota by the passed value then checking if it has been exceeded.
+    */
+  override def isQuotaExceededBy(value: Int): Boolean = {
+    try {
+      sensor.checkQuotaWithDelta(value)
+    } catch {
+      case qve: QuotaViolationException =>
+        logger.info("%s: Quota violated for sensor (%s), metric: (%s), metric-value: (%f), bound: (%f), proposedValue (%s)".format(replicationType, sensor.name(), qve.metricName, qve.value, qve.bound, value))
+        return true
+    }
+    false
+  }
+
+  /**
+    * Returns the bound of the configured quota
+    *
+    * @return
+    */
+  override def bound(): Int = {
```

Comment:
Hmm, should bound() return int? With things like infiniband, it actually can be legit to set a quota larger than 2GB/sec.